src/number_theory/primorial.lean

```diff
@@ -0,0 +1,264 @@
+
+private lemma twice_m_sub_m (m : nat) : 2 * m - m = m :=
+begin
+  rw two_mul m,
+  exact nat.add_sub_cancel m m,
+end
+
+lemma choose_symmetric
+  (m : nat)
+  : nat.choose (2 * m + 1) (m + 1) = nat.choose (2 * m + 1) m
+  :=
+begin
+  apply nat.choose_symm_of_eq_add,
+  ring,
+end
+
+lemma extend_prime_product_odd
+  (n : nat)
+  (n_big : 2 < n)
+  (r : n % 2 = 0)
+  : ∏ i in (finset.filter nat.prime (finset.range n)), i = ∏ i in (finset.filter nat.prime (finset.range (n + 1))), i
+  :=
+begin
+  have not_prime : nat.prime n → false,
+  {
+    intros is_prime,
+    cases (nat.prime.eq_two_or_odd is_prime),
+    { subst h, linarith, },
+    { linarith }
+  },
+  have t :
+    ∏ i in (finset.filter nat.prime (finset.range n.succ)), i
+    = (∏ i in (finset.filter nat.prime (finset.range n)), i) * (if nat.prime n then n else 1),
+  {
+    have s : finset.range n.succ = insert n (finset.range n), by exact finset.range_succ,
+    rw [s, finset.filter_insert, if_neg not_prime, if_neg not_prime],
+    ring,
+  },
+  rw t,
+  simp only [not_prime, mul_one, if_false],
+end
+
+private lemma range_contains_upper_bound (n : nat) : n ∈ finset.range n.succ :=
+begin
+  suffices : (n < n + 1), by simpa,
+  exact lt_add_one n,
+end
+
+lemma select_from_sum (f : nat → nat) (s : finset nat) (a : nat) (t : a ∈ s) : f a ≤ ∑ i in s, f i :=
+begin
+  rw <- finset.insert_erase t,
+  simp,
+end
+
+lemma bound_binomial_coefficient (n : nat) : nat.choose (2 * n + 1) n ≤ 4 ^ n :=
+begin
+  have t : nat.choose (2 * n + 1) n ≤ ∑ i in finset.range (n + 1), nat.choose (2 * n + 1) i,
+    exact select_from_sum (nat.choose (2 * n + 1)) (finset.range (n + 1)) n (range_contains_upper_bound n),
+  simpa [sum_range_choose_halfway n] using t
+end
+
+private lemma can_halve : ∀ (m : nat), (m % 2 = 0) → ∃ n, 2 * n = m
+| 0 := λ _, ⟨ 0, rfl ⟩
+| 1 := λ one_even, by {exfalso, norm_num at one_even}
+| (nat.succ (nat.succ m)) := λ is_even, by {
+  rcases can_halve m is_even with ⟨ half, is_half ⟩,
+  use (half + 1),
+  ring,
+  subst is_half,
+}
+
+private lemma halving_wellfounded (m n : nat) (pr : 2 * m = n + 1) (m_big : 0 < m) : m + 1 < n.succ.succ :=
+begin
+  have u : n.succ.succ = (n + 1) + 1, by norm_num,
+  rw u,
+  rw <- pr,
+  simp only [add_lt_add_iff_right],
+  linarith,
+end
+
+private lemma increment_bit : ∀ (m : nat), (m % 2 = 1) → m.succ % 2 = 0
+| 0 := λ pr, by norm_num at pr
+| 1 := λ pr, by norm_num
+| (nat.succ (nat.succ m)) := λ pr, by
+begin
+  rw nat.mod_def _ 2,
+  simp only [true_and, nat.succ_pos', nat.succ_sub_succ_eq_sub, nat.sub_zero],
+  rw increment_bit m (by exact pr),
+  rw if_pos,
+  exact sup_eq_left.mp rfl,
+end
+
+lemma succ_not_lt (m : nat) : m.succ < m → false :=
+begin
+  intros pr,
+  induction m, {norm_num at pr}, {exact m_ih (nat.lt_of_succ_lt_succ pr)},
+end
+
+lemma middling_prime_divides_binom
+  (p : nat)
+  (is_prime : nat.prime p)
+  (m : nat)
+  (p_big : m.succ < p)
+  (p_small : p ≤ 2 * m + 1)
+  : p ∣ nat.choose (2 * m + 1) m.succ
```

Comment:
I think there is a missing generalization of this: If p > a, p > b but p <= a + b then p divides (a+b) choose a.
`nat.prime.dvd_choose` is another special case, but the general case does not seem to exist.
It looks like the proof of `nat.prime.dvd_choose` should directly generalize.